plugins/org.eclipse.xtext.util/src/org/eclipse/xtext/util/formallang/NfaUtil.java

```diff
@@ -321,13 +318,22 @@ public Object apply(S from) {
 		});
 	}
 
-	public <S> boolean equalsIgnoreOrder(Nfa<S> nfa1, Nfa<S> nfa2, Function<S, ? super Object> keyFunc) {
+	public <S> int hashCode(Nfa<S> nfa, Function<S, ? extends Object> keyFunc) {
+		int result = 1;
+		for (S element : collect(nfa)) {
+			Object s = keyFunc.apply(element);
+			result = 31 * result + (s == null ? 0 : s.hashCode());
```

Comment:
Could be fixed by xoring the hashcodes of all the nfas rather than multiplying with a prime.